PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs

```diff
@@ -0,0 +1,532 @@
+			{
+				if ((content is GeckoHtmlElement) && (!_inFocus))
+				{
+					// The following is required because we get two in focus events every time this
+					// is entered.  This is normal for Gecko.  But I don't want to be constantly
+					// refocussing.
+					_inFocus = true;
+					EnsureFocusedGeckoControlHasInputFocus();
+					if (_browser != null)
+					{
+						_browser.SetInputFocus();
+					}
+					_focusElement = (GeckoHtmlElement)content;
+					ChangeFocus();
+				}
+			}
+		}
+		protected virtual void OnDomDocumentCompleted(object sender, EventArgs ea)
+		{
+			GeckoDocumentCompletedEventArgs e = new GeckoDocumentCompletedEventArgs(ea);
+			_browserDocumentLoaded = true;  // Document loaded once
+			AdjustHeight();
+			if (_entered)
+			{
+				this.Focus();
+
+				IContainerControl containerControl = GetContainerControl();
+
+				if ((containerControl != null) && (containerControl != this) && (containerControl.ActiveControl != this))
+					containerControl.ActiveControl = this;
+
+				_browser.WebBrowserFocus.Activate();
+				var element = (GeckoHtmlElement)_browser.Document.GetElementById("main");
+				element.Focus();
+
+				EnsureFocusedGeckoControlHasInputFocus();
+			}
+		}
+
+		protected override void OnLeave(EventArgs e)
+		{
+			if (_browser != null)
+			{
+				this._browser.RemoveInputFocus();
+			}
+			if ((_browser != null) && (_browser.WebBrowserFocus != null))
+			{
+				_browser.WebBrowserFocus.Deactivate();
+			}
+
+			_entered = false;
+
+			base.OnLeave(e);
+#if __MonoCS__
+			Action EnsureXInputFocusIsRemovedFromReceivedWinFormsControl = () =>
+			{
+				var control = Control.FromHandle(NativeReplacements.MonoGetFocus());
+				var controlName = control.GetType().FullName;
+				if (controlName == "Gecko.GeckoWebBrowser")
+				{
+					return;
+				}
+				MoveInputFocusBacktoAWinFormsControl();
+
+				// Setting the ActiveControl ensure a Focus event occurs on the control focus is moving to.
+				// And this allows us to call RemoveinputFocus at the neccessary time.
+				// This prevents keypress still going to the gecko controls when a winform TextBox has focus
+				// and the mouse is over a gecko control.
+				Form.ActiveForm.ActiveControl = control;
+				EventHandler focusEvent = null;
+				// Attach a execute once only focus handler to the Non GeckoWebBrowser control focus is moving too...
+				focusEvent = (object sender, EventArgs eventArg) =>
+				{
+					control.GotFocus -= focusEvent;
+					MoveInputFocusBacktoAWinFormsControl();
+				};
+
+				control.GotFocus += focusEvent;
+			};
+
+			ProgressUtils.InvokeLaterOnUIThread(() => EnsureXInputFocusIsRemovedFromReceivedWinFormsControl());
+#endif
+		}
+
+		protected void ChangeFocus()
+		{
+			_focusElement.Focus();
+			if (UserGotFocus != null)
+			{
+				UserGotFocus.Invoke(this, null);
+			}
+		}
+
+		protected override void OnEnter(EventArgs e)
+		{
+			_entered = true;
+
+			// Only do this is the control's html has loaded.
+			if (_browserDocumentLoaded)
```

Comment:
Do this only *if* . . .
(or omit the comment entirely, as the if statement is pretty self-explanatory